Add example for using a voter to restrict switch_user

jwmickey opened the pull request.

jwmickey committed:
[WIP] add example for using a voter to restrict switch_user

atailouloute commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,178 @@ setting:
+
+First, create the voter class::
+
+    namespace AppBundle\Security\Voter;
```
`namespace App\Security\Voter;`to be coherent with SF4 structure

atailouloute commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,178 @@ setting:
+
+            if (in_array('ROLE_CUSTOMER', $subject->getRoles()) 
+                && $this->hasSwitchToCustomerRole($token)) {
+                return self::ACCESS_GRANTED;
```
It should be `return true;` no?

atailouloute commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,178 @@ setting:
+
+        // Explicitly configure the service
+        $container->getDefinition(SwitchToCustomerVoter::class)
+            ->setArgument('$roleHierarchy', '@security.role_hierarchy');   
```
This should be `new Reference('security.role_hierarchy')`

jwmickey committed:
Merge branch 'master' of https://github.com/symfony/symfony-docs into fix_9095

jwmickey committed:
Fix recommendations from community review

jwmickey commented:
Thank you for reviewing this PR @atailouloute, I have added your recommendations.

xabbuh commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,179 @@ setting:
+
+        private function hasSwitchToCustomerRole(TokenInterface $token)
+        {
+            $roles = $this->roleHierarchy->getReachableRoles($token->getRoles());
```
I would leave out the role hierarchy here. It IMO adds more complexity than needed to understand the example.

xabbuh commented:
I think we should add a headline before the new section to make it stand out more. And we should then also add a `versionadded` directive.

ndench reviewed (approved):
Thanks for the contribution. This is a cool idea, at my old job we just gave ROLE_SUPER_ADMIN to all the employees, which always felt gross. If only we'd thought of doing this!

iltar commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,179 @@ setting:
         ));
 
+If you need more control over user switching, but don't require the complexity 
+of a full ACL implementation, you can use a security voter.  For example, you 
```
2 spaces should be one `voter.  For example`

iltar commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,179 @@ setting:
+
+    Notice that when checking for the ``ROLE_CUSTOMER`` role on the target user, only the roles 
+    explicitly assigned to the user are checked rather than checking all reachable roles from
+    the role hierarchy.  The reason for this is to avoid accidentally granting access to an 
```
double spaces: `hierarchy.  The reason`

iltar commented on security/impersonating_user.rst:
```diff
@@ -187,6 +187,179 @@ setting:
+            ),
+        ));
+
+Thanks to autowiring, we only need to configure the role hierarchy argument when registering 
```
`we only need to configure`. Correct me if I'm wrong, but I believe the Symfony docs point towards "you", rather than "we".

ndench reviewed (approved):
Sweet example.

jwmickey committed:
implemented community suggestions

- removed role_heirarchy from switch user voter to simplify
- removed double-spaces between sentences.
- added versionadded
- added header

javiereguiluz commented:
@jwmickey thanks a lot for contributing this nice example (and thanks to all reviewers!). I apologize for having taken so long to merge it.

I tried to merge but there were lots of conflicts ... so I created #10842 to replace this and that is now merged. Of course I kept your original commits so you'll get full credit of your contribution.

So, let's close this now that #10842 is merged. Thanks!

jwmickey commented:
Thanks @javiereguiluz, glad this was able to be included!
